sdk/webpubsub/Microsoft.Azure.WebJobs.Extensions.WebPubSub/src/Bindings/WebPubSubOperation.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Runtime.Serialization;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Converters;
+
+namespace Microsoft.Azure.WebJobs.Extensions.WebPubSub
+{
+    /// <summary>
+    /// Supported operations of rest calls.
+    /// </summary>
+    [JsonConverter(typeof(StringEnumConverter))]
+    public enum WebPubSubOperation
```

Comment:
This is not something I've seen before while working in the messaging extensions. Generally when this much flexibility is required, we allow binding to the service client rather than defining an enum that has all possible operations.

Reply:
Right. This is because our rest call supports different actions and require different parameters. Binding to service client is not supported in non-csharp language. And customer will need to build their own client by reference our server sdk to do this. Then output binding is not needed anymore. To provide a consistent function binding experience, this is the best way we can cover all rest calls for now. How do you think?

Reply:
I think I understand, but we have this same consideration for all extensions. I guess the difference might be that in other extensions the behavior can be configured statically through the output binding attribute. Is that something that would work here? i.e. you would define the Operation at the function level rather than at the function invocation level.

Reply:
Make it in attribute will lead to output binding is limited to do a single operation. I think customer need to trigger a few output bindings in some cases.
For example: 
```cs
[FunctionName("connected")]
public static async Task Connected(
    [WebPubSubTrigger(WebPubSubEventType.System, "connected")] ConnectionContext connectionContext,
    [WebPubSub] IAsyncCollector<WebPubSubEvent> webpubsubEvent)
{
    await webpubsubEvent.AddAsync(new WebPubSubEvent
    {
        Operation = WebPubSubOperation.SendToAll,
        Message = new WebPubSubMessage(new ClientContent($"{connectionContext.UserId} connected.").ToString()),
        DataType = MessageDataType.Json
    });

    await webpubsubEvent.AddAsync(new WebPubSubEvent
    {
        Operation = WebPubSubOperation.AddUserToGroup,
        UserId = connectionContext.UserId,
        Group = "group1"
    });
    await webpubsubEvent.AddAsync(new WebPubSubEvent
    {
        Operation = WebPubSubOperation.SendToUser,
        UserId = connectionContext.UserId,
        Group = "group1",
        Message = new WebPubSubMessage(new ClientContent($"{connectionContext.UserId} joined group: group1.").ToString()),
        DataType = MessageDataType.Json
    });
}
```

Reply:
Still, this could be configured statically by allowing multiple operations to be specified in the attribute with a Flags enum. The point where this falls apart would be when users need to be able to inspect the event to figure out what operation to perform. If this can be considered an advanced scenario, then it may be worth considering allowing the operation to be selected statically, and having the user just supply the WebPubSubEvent. The event then becomes more of a typical data transfer object.

Reply:
Some advanced functions in js may be used like below. However, if limited to attributes, some customized property is not able to use.

```js
module.exports = function (context, connectionContext) {
  context.bindings.webPubSubEvent = [];
  context.bindings.webPubSubEvent.push({
    "operation": "sendToAll",
    "message": JSON.stringify({
        from: '[System]',
        content: `${context.bindingData.connectionContext.userId} connected.`
      }),
    "dataType" : "json"
  });

  context.bindings.webPubSubEvent.push({
    "operation": "addUserToGroup",
    "userId": `${context.bindingData.connectionContext.userId}`,
    "group": "group1"
  });

  context.bindings.webPubSubEvent.push({
    "operation": "sendToAll",
    "message": JSON.stringify({
          from: '[System]',
          content: `${context.bindingData.connectionContext.userId} joined group: group1.`
      }),
    "dataType": "json"
  });
  context.done();
};
```

Maybe the naming leads confusing about this output binding usage. I'm rename to `WebPubSubOperation` as a operation  triggered from server to ask service to manage clients. Making this in function extension is to make it extensible as we may support persistent connections in the future and then the operations is not limited to current server SDK. (Probably in another sdk)

I'm splitting this operation in different classes so customer would be clear of what parameters are needed and what operation is triggered clearly. Do you think this is better?

```cs
[FunctionName("connected")]
public static async Task Connected(
    [WebPubSubTrigger(WebPubSubEventType.System, "connected")] ConnectionContext connectionContext,
    [WebPubSub] IAsyncCollector<WebPubSubOperation> webpubsubOperations)
{
    await webpubsubOperations.AddAsync(new SendToAll
    {
        Message = BinaryData.FromString(new ClientContent($"{connectionContext.UserId} connected.").ToString()),
        DataType = MessageDataType.Json
    });

    await webpubsubOperations.AddAsync(new AddUserToGroup
    {
        UserId = connectionContext.UserId,
        Group = "group1"
    });
    await webpubsubOperations.AddAsync(new SendToUser
    {
        UserId = connectionContext.UserId,
        Message = BinaryData.FromString(new ClientContent($"{connectionContext.UserId} joined group: group1.").ToString()),
        DataType = MessageDataType.Json
    });
```

Reply:
Yes, I do think that the semantics here are easier to understand. It also makes it clearer which parameters go with which operation.